File: BotsSDK/korebot/src/main/java/kore/botssdk/adapter/ListWidgetButtonAdapter.java
package kore.botssdk.adapter;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.PopupWindow;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.RequestOptions;
import com.google.android.material.bottomsheet.BottomSheetDialog;

import java.util.ArrayList;

import kore.botssdk.R;
import kore.botssdk.listener.ComposeFooterInterface;
import kore.botssdk.listener.InvokeGenericWebViewInterface;
import kore.botssdk.models.BotResponse;
import kore.botssdk.models.Widget;
import kore.botssdk.net.SDKConfiguration;
import kore.botssdk.utils.BundleConstants;
import kore.botssdk.viewholders.BaseViewHolder;

public class ListWidgetButtonAdapter extends RecyclerView.Adapter<ListWidgetButtonAdapter.ButtonViewHolder> {
    private final LayoutInflater inflater;
    private final ArrayList<Widget.Button> buttons;
    private ComposeFooterInterface composeFooterInterface;
    private InvokeGenericWebViewInterface invokeGenericWebViewInterface;
    private String skillName;
    private BottomSheetDialog bottomSheetDialog;
    private PopupWindow popupWindow;
    int displayLimit = -1;
    private boolean isEnabled = true;
    private int type;
    private boolean isMenus = false;

    public ListWidgetButtonAdapter(Context context, ArrayList<Widget.Button> buttons, boolean isMenus) {
        this.buttons = buttons;
        this.inflater = LayoutInflater.from(context);
        this.isMenus = isMenus;
    }

    @NonNull
    @Override
    public ButtonViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        if (type == 1) {
            return new ButtonViewHolder(inflater.inflate(R.layout.advance_button_fullwidth, viewGroup, false));
        }
        return new ButtonViewHolder(inflater.inflate(R.layout.widget_button_list_item, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ButtonViewHolder holder, int i) {
        Widget.Button btn = buttons.get(i);
        holder.tv.setText(btn.getTitle());
        holder.tv.setTextColor(Color.parseColor(SDKConfiguration.BubbleColors.quickReplyColor));
        Drawable errorDrawable = BaseViewHolder.getTintDrawable(holder.tv.getContext(), SDKConfiguration.BubbleColors.quickReplyColor, R.drawable.check_icon);
        holder.image.setVisibility(isMenus ? GONE : VISIBLE);
        if (!isMenus) {
            if (btn.getImage() != null && btn.getImage().getImage_type().equals(BotResponse.COMPONENT_TYPE_IMAGE) && !btn.getImage().getImage_src().isEmpty()) {
                Glide.with(holder.itemView.getContext())
                        .load(btn.getImage().getUrl() != null ? btn.getImage().getUrl() : btn.getImage().getImage_src())
                        .error(errorDrawable)
                        .apply(new RequestOptions().diskCacheStrategy(DiskCacheStrategy.NONE))
                        .into(holder.image);
            } else {
                holder.image.setImageDrawable(errorDrawable);
            }
        }

        ((ViewGroup) holder.tv.getParent()).setOnClickListener(v -> {
            if (bottomSheetDialog != null) bottomSheetDialog.dismiss();
            if (popupWindow != null && popupWindow.isShowing()) popupWindow.dismiss();

            if (invokeGenericWebViewInterface != null) {
                if (BundleConstants.BUTTON_TYPE_WEB_URL.equalsIgnoreCase(btn.getType())) {
                    invokeGenericWebViewInterface.invokeGenericWebView(btn.getUrl());
                    return;
                } else if (BundleConstants.BUTTON_TYPE_URL.equalsIgnoreCase(btn.getType())) {
                    invokeGenericWebViewInterface.invokeGenericWebView(btn.getUrl());
                    return;
                }
            }
            if (composeFooterInterface != null && isEnabled) {
                composeFooterInterface.onSendClick(btn.getTitle(), btn.getPayload(), false);
            }
        });
    }

    public void setDisplayLimit(int displayLimit) {
        this.displayLimit = displayLimit;
    }

    public void setType(int type) {
        this.type = type;
    }

    public void setComposeFooterInterface(ComposeFooterInterface composeFooterInterface) {
        this.composeFooterInterface = composeFooterInterface;
    }

    public void setInvokeGenericWebViewInterface(InvokeGenericWebViewInterface invokeGenericWebViewInterface) {
        this.invokeGenericWebViewInterface = invokeGenericWebViewInterface;
    }

    public void setEnabled(boolean enabled) {
        isEnabled = enabled;
    }

    public void setBottomSheet(BottomSheetDialog bottomSheetDialog) {
        this.bottomSheetDialog = bottomSheetDialog;
    }

    public void setPopupWindow(PopupWindow popupWindow) {
        this.popupWindow = popupWindow;
    }

    @Override
    public int getItemCount() {
        if (buttons != null) {
            return (displayLimit > -1 && buttons.size() > displayLimit) ? displayLimit : buttons.size();
        }

        return 0;
    }

    boolean isFullView;

    public void setIsFromFullView(boolean isFullView) {
        this.isFullView = isFullView;
    }

    public static class ButtonViewHolder extends RecyclerView.ViewHolder {
        final TextView tv;
        final ImageView image;

        public ButtonViewHolder(@NonNull View itemView) {
            super(itemView);
            tv = itemView.findViewById(R.id.buttonTV);
            image = itemView.findViewById(R.id.ivBtnImage);
        }
    }

    public String getSkillName() {
        return skillName;
    }

    public void setSkillName(String skillName) {
        this.skillName = skillName;
    }

    public static void showEmailIntent(Activity activity, String recipientEmail) {
        Intent emailIntent = new Intent(Intent.ACTION_SENDTO);
        emailIntent.setData(Uri.parse("mailto:" + recipientEmail));
        emailIntent.putExtra(Intent.EXTRA_SUBJECT, "");

        try {
            activity.startActivity(emailIntent);
        } catch (ActivityNotFoundException e) {
            Toast.makeText(activity, "Error while launching email intent!", Toast.LENGTH_SHORT).show();
        }
    }
}

File: BotSdk/ui/src/main/java/com/kore/ui/adapters/ListWidgetButtonAdapter.kt
package com.kore.ui.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.PopupWindow
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.graphics.toColorInt
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.target.DrawableImageViewTarget
import com.kore.common.event.UserActionEvent
import com.kore.data.repository.preference.PreferenceRepositoryImpl
import com.kore.event.BotChatEvent
import com.kore.model.constants.BotResponseConstants
import com.kore.model.constants.BotResponseConstants.BUBBLE_RIGHT_BG_COLOR
import com.kore.model.constants.BotResponseConstants.THEME_NAME
import com.kore.ui.R

class ListWidgetButtonAdapter(
    val context: Context,
    private val buttons: List<Map<String, *>>,
    private val displayLimit: Int,
    private val style: String,
    private val popupWindow: PopupWindow?,
    private val isLastItem: Boolean,
    private var isMenus: Boolean = false,
    private val actionEvent: (event: UserActionEvent) -> Unit
) : RecyclerView.Adapter<ListWidgetButtonAdapter.ButtonViewHolder>() {

    class ButtonViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val tvButtonTv: TextView = itemView.findViewById(R.id.buttonTV)
        val ivBtnImage: ImageView = itemView.findViewById(R.id.ivBtnImage)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ButtonViewHolder {
        return if (style == BotResponseConstants.FIT_TO_WIDTH) ButtonViewHolder(
            LayoutInflater.from(context).inflate(R.layout.widget_button_list_full_item, parent, false)
        )
        else ButtonViewHolder(LayoutInflater.from(context).inflate(R.layout.widget_button_list_item, parent, false))
    }

    override fun onBindViewHolder(holder: ButtonViewHolder, position: Int) {
        val btn = buttons[position]
        val image = btn[BotResponseConstants.COMPONENT_TYPE_IMAGE] as Map<String, *>
        val context = holder.itemView.context
        val sharedPrefs = PreferenceRepositoryImpl()
        val txtColor = sharedPrefs.getStringValue(context, THEME_NAME, BUBBLE_RIGHT_BG_COLOR, "#3F51B5").toColorInt()
        holder.tvButtonTv.setTextColor(txtColor)
        holder.tvButtonTv.text = btn[BotResponseConstants.KEY_TITLE] as String
        holder.ivBtnImage.isVisible = !isMenus

        holder.tvButtonTv.setOnClickListener {
            popupWindow?.dismiss()
            if (BotResponseConstants.BUTTON_TYPE_WEB_URL == btn[BotResponseConstants.TYPE] ||
                BotResponseConstants.BUTTON_TYPE_URL == btn[BotResponseConstants.TYPE]
            ) {
                actionEvent(BotChatEvent.UrlClick(btn[BotResponseConstants.URL].toString()))
            } else if (isLastItem && btn[BotResponseConstants.PAYLOAD] != null) {
                actionEvent(BotChatEvent.SendMessage(holder.tvButtonTv.text.toString(), btn[BotResponseConstants.PAYLOAD] as String))
            }
        }
        val errorDrawable = ContextCompat.getDrawable(context, R.drawable.ic_correct)
        errorDrawable?.setTint(txtColor)
        if (image[BotResponseConstants.IMAGE_SRC] != null && !(image[BotResponseConstants.IMAGE_SRC] as String?).isNullOrEmpty()) {
            Glide.with(context).load(image[BotResponseConstants.IMAGE_SRC]).error(errorDrawable)
                .into<DrawableImageViewTarget>(DrawableImageViewTarget(holder.ivBtnImage))
        } else holder.ivBtnImage.setImageDrawable(errorDrawable)
    }

    override fun getItemCount(): Int {
        return displayLimit
    }
}